### ISSUE TITLE
Question about calculation of the IGD - InvertedGenerationalDistance

### ISSUE THREAD

**Issue by saviosampaio** [MOEAFramework/MOEAFramework#113]
Hi David Hadka,

I am reading this documents:

[1] Hadka, David, "Beginner's Guide to the MOEAFramework", 2016.

[2] Q. Zhang, A. Zhao, P. N. Suganthan, W. Liu, and S. Tiwari, "Multi-objective optimization test instances for the CEC09 special session and competition", University of Essex and Nanyang Technological University, Technical Report CES-487, 2008

There is, on this document [2], a item: "IV. PERFORMANCE ASSESSMENTS", "A. Performance Metric (IGD)".

In document [1] we can read: "Inverted generational distance is the average distance from every solution in the reference set to the nearest solution in the approximation set.". This document [1] quotes the author Coello (Coello Coello et al., 2007).

I would like to know if the "InvertedGenerationalDistance" calculated in MOEAFramework uses all individuals from the last population, or if MOEAFramework uses only NonDominated individuals from the last population.

This is because when we use the "analyzer.saveData" option, the saved files do not display all individuals of each generation. 

Example: I used populationSize=100, and saved file from "analyzer.saveData" option shows 15 or 30 individuals between the character "#".

I also thought that the "#" character would be used to separate the "dominance rank", but the saved file contains more than 500 lines, number greater than the population size.

I am afraid that the IGD is being calculated only with part of the population and that this may cause an interpretation different from that presented in the document [2].

Please, could you explain if the IGD calculation used in the MOEAFramework is in accordance with the CEC2009 document?

Thank you very much.

Sávio Sampaio

**Comment by dhadka** [comment]
Hi, thanks for the question.

I believe the implementation matches what is described in [2].  They state "Let A be an approximation set to the PF".

From Coello Coello's [Evolutionary Algorithms for Solving Multi-Objective Problems book](https://books.google.com/books?id=rXIuAMw3lGAC&lpg=PA779&ots=-j2-ZT4MRK&dq=Evolutionary%20Algorithms%20for%20Solving%20Multi-Objective%20Problems%20inverted%20generational%20distance&pg=PA251#v=onepage&q=Evolutionary%20Algorithms%20for%20Solving%20Multi-Objective%20Problems%20inverted%20generational%20distance&f=false), the approximation set is defined as the set of vectors that do not weakly dominate any other vector in the set.

Thus, the calculation for IGD is comparing only the non-dominated solutions in the final population to the Pareto front.

The # you see in the output file separates approximation sets.  That is, if you run 50 different seeds, the file will contain 50 different approximation sets, each separated by the # character.

Hope that helps answer your question!

**Comment by saviosampaio** [comment]
Hi David Hadka,

Thank you by your quick answer and attention.

Now I understand better the character "#" in saved data.

My question is because there are cases where the final population (last generation) from some algorithm may not contain 100 solutions ("populationSize"), for example, displaying only 80 non-dominated solutions in their latest generation. And so, I figured this could misrepresent the IGD calculation result, compared to another algorithm that featured 100 solutions in its latest generation.

I was also surprised that in a simple test, with the algorithm **"NSGAII", using "operator" "de+pm"**, and the "UF13" problem, the IGD value was **approximately 1.3089**, and using your default "operator" "sbx+pm" the IGD value of "NSGAII" was **approximately 1.47226**, when the best CEC2009 result for "UF13" problem was **1.8489 (MOEAD)**, and other results were much larger, such as **2.2057 (GDE3) and 3.2323 (NSGAIILS)**, among others.

However, I have verified that the saved file (with analyzer.saveData) contains 30 results (separated by the "#" character), with 100 solutions each.

Your response was very helpful.

I believe that this "question issue" can be closed.

Thank you very much, again.

Sávio Sampaio
